Stack and concatenate

andrei-papou opened the pull request:
Issue: https://github.com/rust-ndarray/ndarray/issues/667

Renamed `stack` to `concatenate` to follow the behaviour of `numpy` and also added `stack` function. Examples:

##### concatenate
```rust
use ndarray::{arr2, Axis, concatenate};

let a = arr2(&[[2., 2.], [3., 3.]]);
assert!(
     concatenate(Axis(0), &[a.view(), a.view()])
     == Ok(arr2(&[[2., 2.], [3., 3.], [2., 2.], [3., 3.]]))
);
```

##### stack
```rust
use ndarray::{arr2, arr3, stack, Axis};

let a = arr2(&[[2., 2.],
               [3., 3.]]);
assert!(
    stack![Axis(0), a, a]
    == arr3(&[[[2., 2.],
               [3., 3.]],
              [[2., 2.],
               [3., 3.]]])
);
```

bluss commented:
Nice, this is the right direction. I think it would be best to make this in the shape of  a non breaking change to start with so that it can be included soon.

bluss commented on src/stacking.rs:
```diff
@@ -76,26 +73,45 @@ where
+    }
+    let arrays: Vec<ArrayView<A, D::Larger>> =
+        arrays.into_iter().map(|a| a.insert_axis(axis)).collect();
+    concatenate(axis, &arrays)
```
Implementation strategy wise, can't we take the old `stack`, put that into an internal function. Give it a parameter that tells it whether to remove an axis or not. It can be a type parameter, since it influences the return type (the dim of the returned array).

That way both stack and concatenate can use this internal function and we don't need the intermediate vector here in concatenate

andrei-papou replied:
Looks good! I'll try to implement it this way.

bluss commented:
This branch will eventually need to be rebased to skip the merges in the history (see commit view). The blast back to #1, #5 etc in ndarray history is fun, but we don't want to have those merges there.

Also, "this is the right direction" isn't supposed to sound like "nice try". It's supposed to say.. This is the right **destination**, this is how we want to name those functions. To be pragmatic I think we'd like to have this in 0.13.x, and then as a non-breaking change in the first round (preparing to put the functions in their right places in the next version).

andrei-papou commented:
@bluss sorry for the delay, the month is pretty busy. I've refactored the new function to get rid of the redundant allocation. Also I renamed the functions in the backward compatible way. Thanks.

LukeMathWalker commented on src/stacking.rs:
```diff
@@ -76,14 +73,57 @@ where
 /// [1]: fn.stack.html
 ///
-/// ***Panics*** if the `stack` function would return an error.
+/// ***Panics*** if the `concatenate` function would return an error.
```
`concatenate` does not exist yet, we shouldn't refer to it here.

LukeMathWalker commented:
If we want to introduce this in `0.13.x`, we should probably:
- mark the existing `stack` function as deprecated, suggesting to use `stack_new_axis`;
- introduce `concatenate` and `concatenate!`, which underneath are just calling `stack` for now;
- introduce `stack_new_axis` which will become the "new" stack.

What do you think? @andrei-papou 
Always difficult to change what a name means :/

andrei-papou commented:
> * mark the existing `stack` function as deprecated, suggesting to use `stack_new_axis`;

@LukeMathWalker shouldn't we suggest to use `concatenate` instead? `concatenate` is a new name for the current `stack` function while `stack_new_axis` behaves in a different way.

LukeMathWalker commented:
Yes, you are right :sweat_smile:

andrei-papou commented:
@LukeMathWalker ready, but I had to suppress a few deprecation warnings: we still need `stack` function in `concatenate` implementation, in `lib.rs` and in tests.

andrei-papou commented:
@bluss @LukeMathWalker could you please do another round of review? Looks like all the comments have been resolved quite a while ago. Thanks.

bluss commented on src/stacking.rs:
```diff
@@ -76,7 +77,79 @@ where
+
+pub fn stack_new_axis<A, D>(
+    axis: Axis,
+    arrays: Vec<ArrayView<A, D>>,
```
`stack_new_axis` and `concatenate` should agree on if they take the arrays as a slice or a Vec, they should have the same interface if they can. Avoiding cloning of the array views seems like a minor issue, but it can be decided either way (slice or vec)

bluss commented on src/stacking.rs:
```diff
@@ -76,7 +77,79 @@ where
+    stack(axis, arrays)
+}
+
+pub fn stack_new_axis<A, D>(
```
Missing doc comment :)

bluss commented on src/stacking.rs:
```diff
@@ -109,3 +182,71 @@ macro_rules! stack {
         $crate::stack($axis, &[ $($crate::ArrayView::from(&$array) ),* ]).unwrap()
```
The stack macro should also be deprecated

bluss commented on src/stacking.rs:
```diff
@@ -76,7 +77,79 @@ where
+    res_dim.set_axis(axis, arrays.len());
+
+    // we can safely use uninitialized values here because they are Copy
+    // and we will only ever write to them
```
This looks good, it follows what we do in stack. We'll revise handling of uninit data, but it's not in scope for this PR. (#796 )

bluss commented on src/stacking.rs:
```diff
@@ -76,7 +77,79 @@ where
+    let mut res = Array::from_shape_vec(res_dim, v)?;
+
+    res.axis_iter_mut(axis)
+        .zip(arrays.into_iter())
```
this zip seems to show that we would be fine with just using a slice of `ArrayView` as input, we don't need to use them by value

bluss commented on src/slice.rs:
```diff
@@ -544,62 +544,62 @@ impl_slicenextdim_larger!((), Slice);
 macro_rules! s(
     // convert a..b;c into @convert(a..b, c), final item
-    (@parse $dim:expr, [$($stack:tt)*] $r:expr;$s:expr) => {
+    (@parse $dim:expr, [$($concatenate:tt)*] $r:expr;$s:expr) => {
```
This rename of the variable inside the macro is an incidental mistake, and should not be in the PR

bluss replied:
Oh I removed this in my rebase-fix of the branch. No other changes in that branch. :)

bluss commented:
I have rebased your commit history and pushed to a new branch: See https://github.com/rust-ndarray/ndarray/pull/new/ap/stack-and-concatenate

Please use it as a starting point - it removes all the merges in the commit history of this PR. We accept some unclean histories, but this needed to be cleaned up before merging.

There are some commits that have the title "Changelog:" please edit those, and make sure the first line is a summary of what the commit does. See my commits in the repo, if you want an example. Thanks :)
